### PR TITLE
Update architecture spec & misc updates

### PR DESCRIPTION
**Pull request by justmoon**
Resolves https://github.com/interledger/rfcs/issues/146

### CODE REVIEW

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 1]
```diff
@@ -194,39 +199,13 @@ Transport layer protocols are responsible for coordinating the different transfe
...
+#### Privately Shared Key (PSK)
 
-##### As a Sub-Protocol
+<span class="show alert alert-warning">**TODO:** Describe PSK.</span>
 
-ATP can be used as a sub-protocol for part of a UTP transaction. This happens when a connector decides to create an ATP transfer as the next hop. It takes on both the risk of being unable to pass on the receipt and the risk of the notaries failing. This makes sense to do if the next connector quotes a better price for an ATP transfer than for a UTP transfer.
+#### Interledger Payment Request (IPR)
 
-Similarly, an ATP transaction can turn into a UTP transaction when a connector decides to use UTP for the next hop. It takes on the risk of not being able to pass on the receipt to the notaries. This makes sense if ATP is not available.
+<span class="show alert alert-warning">**TODO:** Describe IPR.</span>
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Shouldn't we just link to the IPR RFC?

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 2]
```diff
@@ -98,47 +100,46 @@ Ledger protocols are responsible for executing the individual transfers that con
...
+* Prepared -- Funds are held.
 * Executed -- The transfer has completed.
 * Rejected -- The transfer has been canceled (and funds returned to the sender.)
 
-<span class="show alert alert-info">**Hint:** Escrow is the financial equivalent of a [two-phase commit](http://foldoc.org/two-phase%20commit).</span>
+<span class="show alert alert-info">**Hint:** Authorization holds are the financial equivalent of a [two-phase commit](http://foldoc.org/two-phase%20commit).</span>
 
-
-The ledger MUST be able to verify a simple SHA-256 hashlock cryptographic escrow condition in order to automatically trigger the release of the escrow. It must also allow a timeout to automatically roll back an escrowed transfer.
+The ledger MUST be able to release the held funds to the receiver upon receiving a 32-byte preimage whose SHA-256 hash matches a value provided by the sender. It MUST also support releasing the funds back to the sender after a timeout.
```
Comment (sharafian) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: For a ledger to be supported, it should accept ONLY 32-byte pre-images. Otherwise a connector could stand to lose money, if they have an outgoing transfer on a ledger with a larger fulfillment size than the ledger of the corresponding incoming transfer.

Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: I'd say that it can accept more (in general) but it needs a way to ensure that for a given transfer the fulfillment will only be considered valid if it is exactly 32 bytes (e.g. like the cost limit in Crypto Conditions)

Comment (sharafian) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: fair enough. rather than saying the ledger supports only 32-byte preimages, we should say you need a way to construct a transfer that can only be fulfilled by a 32-byte preimage

Comment (sharafian) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: is it possible to do that on bitcoin with some kind of bitcoin scripts magic?

Comment (stevenroose) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: @sharafian yeah. It's basically an HTLC like Lightning uses them. Except it's not from a two-party escrow address but just a one-to-one HTLC.

Comment (adrianhopebailie) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: @sharafian if we want to leave the door open for other sizes then we should say that the ledger must allow a client to specify the expected size of the fulfillment when requesting the transfer. If that is not provided it must be assumed to be 32 bytes.

File: 0001-interledger-architecture/0001-interledger-architecture.md (hunk 1 above)
Comment (justmoon) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: We'll definitely link it, but the previous version of this doc had a short paragraph for each protocol / component, which was quite useful I thought.

File: 0001-interledger-architecture/0001-interledger-architecture.md (hunk 2 above)
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: @stevenroose I know Lightning uses HTLCs with 32-byte preimages, but do you know if the scripts they use _guarantee_ that the preimage could not be longer than 32 bytes?

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 3]
```diff
@@ -98,47 +100,46 @@ Ledger protocols are responsible for executing the individual transfers that con
...
 
 The ledger MUST support attaching a short message or *memo* to each transfer.
 
 #### Full Support
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Should we mention speed here? If you want to fully support ILP the ledger should be able to prepare and execute transfers in ~1 second or less.

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 4]
```diff
@@ -98,47 +100,46 @@ Ledger protocols are responsible for executing the individual transfers that con
 
 #### Minimal Support
 
-For minimal Interledger support, the ledger MUST have the ability for funds to be transferred from one account to another. This enables OTP transactions to pass through this ledger. It also allows third parties to act as escrow providers in order to enable UTP and ATP transactions on the ledger.
+For minimal Interledger support, the ledger MUST have the ability for funds to be transferred from one account to another. All of the more advanced ledger layer functionality, such as escrowed transfers, messaging, etc. can be implemented on top of this type of basic ledger.
 
 #### Basic Support
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: What about mentioning notifications?

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 5]
```diff
@@ -98,47 +100,46 @@ Ledger protocols are responsible for executing the individual transfers that con
...
-It MUST support all cryptographic condition types with the status "recommended".
-
-It MUST support memos to be added for the credited and debited accounts separately and SHOULD support fairly large memo sizes.
+It MUST support memos up to 65535 bytes.
 
-It MUST support a way to discover connectors connected to the ledger.
+It MUST support sending an authenticated message of up to 65535 bytes to the holder of another account on the ledger.
 
-It MUST support a way to look up a fulfillment by condition hash. It SHOULD automatically reject new transfers (that have not been prepared yet) that have an execution condition for which the ledger already knows the fulfillment. This aids in [UTP error recovery](#error-recovery).
+It MUST support a way to look up a fulfillment by condition hash. It SHOULD automatically reject new transfers (that have not been prepared yet) that have an execution condition for which the ledger already knows the fulfillment. This aids in [error recovery](#error-recovery).
```
Comment (adrianhopebailie) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: We should mention that it is necessary for a ledger protocol to define an addressing scheme that complies with the ILP Addressing scheme such that accounts in the ledger can be addressed using ILP Addresses.

See: https://github.com/interledger/rfcs/issues/31#issuecomment-286469564

File: 0001-interledger-architecture/0001-interledger-architecture.md (hunk 3 above)
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: What about enabling off-ledger, ledger-secured payment channels? That might be for more-than-full support

File: 0003-interledger-protocol/0003-interledger-protocol.md [hunk 6]
```diff
@@ -172,67 +131,66 @@ As with the [internet protocol](https://tools.ietf.org/html/rfc791#section-2.3),
...
-| destinationAddress | IlpAddress | Address corresponding to the destination account |
-| destinationAmount | IlpAmount | Amount the destination account should receive, denominated in the asset of the destination ledger |
-| condition | IlpCondition | Execution condition for the transfers
-| expiresAt | IlpTimestamp | Maximum expiry time of the last transfer that the recipient will accept |
-
-#### version
-
-    INTEGER(0..255)
+| amount | UInt64 | Amount the destination account should receive, denominated in the asset of the destination ledger |
+| account | Address | Address corresponding to the destination account |
```
Comment (michielbdejong) [inline 0003-interledger-protocol/0003-interledger-protocol.md]: Why the change from 'address' to 'account'? Will this same change be made in the [IPR](https://github.com/interledger/rfcs/blob/master/0011-interledger-payment-request/0011-interledger-payment-request.md#json) format?

Comment (sharafian) [inline 0003-interledger-protocol/0003-interledger-protocol.md]: IPR is (probably) going to use the data format from the new version of PSK once we switch to the binary ILP packet format.

Paraphrasing @justmoon 's reasoning on `account` vs. `address`, the account is the actual thing being represented, while an address is merely the data format of it (like calling the amount "number")

**Comment by mDuo13** [comment]
From https://github.com/interledger/java-ilp-core/pull/18#discussion_r106468355 —

It's confusing that ILP, or the Interledger Protocol (Interledger Project?), is actually a protocol suite containing a protocol specifically named "Interledger Protocol" (IL-RFC-3). We should strongly consider renaming IL-RFC-3 to make this less confusing. I don't have any good candidates for what else to call it, though.

**Comment by adrianhopebailie** [comment]
> It's confusing that ILP, or the Interledger Protocol (Interledger Project?), is actually a protocol suite containing a protocol specifically named "Interledger Protocol" (IL-RFC-3)

This RFC is called the "Interledger architecture" (i.e. not the Interledger Protocol architecture). Maybe the problem is that we use the acronym ILP too liberally when it should only be used to describe the Interledger Protocol (i.e. IL-RFC-3).

i.e. The Interledger Architecture defines a number of protocol layers including the Interledger Protocol (ILP) layer.

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 7]
```diff
@@ -98,47 +100,46 @@ Ledger protocols are responsible for executing the individual transfers that con
...
 
-The ledger MUST provide cryptographic escrow. Escrow means that transfers can have four states:
+The ledger MUST provide cryptographic authorization holds. During an authorization hold, money is put aside for a specific transfer until that transfer's outcome has been decided.
+
+Transfers using authorization holds can be in four distinct states:
 
-![enter image description here](https://lh3.googleusercontent.com/-QHsehrPTXgM/Vu_zs1dLWPI/AAAAAAAAEys/zchHKJj_MQs/s0/Transfer%252520States.png "Transfer States")
+![Transfers start in the proposed state, transition to the prepared state and finally to the executed state, which is final. Proposed and prepared transfers can also transition to the rejected state, which is final as well.](../shared/graphs/transfer-states.svg)
 
 * Proposed -- Nothing has happened yet.
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Can we take out the `Proposed` state?

File: 0001-interledger-architecture/0001-interledger-architecture.md (hunk 3 above)
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Also, what about account notifications?

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 8]
```diff
@@ -98,47 +100,46 @@ Ledger protocols are responsible for executing the individual transfers that con
...
-#### Simple Ledger Protocol (SLP)
+#### Five Bells Ledger Protocol (5BLP)
 
-Simple Ledger Protocol (SLP) is a RESTful, JSON-based protocol that was developed specifically to provide the minimum functionality required for full Interledger support.
+Five Bells Ledger Protocol (5BLP) is a RESTful, JSON-based protocol that was developed specifically to provide the minimum functionality required for full Interledger support.
 
-A reference implementation of a ledger using SLP can be found [here](https://github.com/interledger/five-bells-ledger).
+A reference implementation of a ledger using 5BLP can be found [here](https://github.com/interledger/five-bells-ledger).
 
 #### Blockchain Protocols (e.g. Bitcoin)
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Remove old mention of OTP/UTP in the bitcoin section

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 9]
```diff
@@ -48,7 +48,7 @@ A basic objective of the Interledger design is to tolerate a wide range of ledge
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: This spec should no longer say "**This is a strawman proposal.** Choices made in this document are merely intended as a starting point." at the top

File: 0001-interledger-architecture/0001-interledger-architecture.md (hunk 7 above)
Comment (justmoon) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Agree that we can remove `Proposed`, but in its place we should add a name for the initial state, i.e. where the ledger does not know about the transfer yet.

File: 0001-interledger-architecture/0001-interledger-architecture.md (hunk 2 above)
Comment (justmoon) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: > do you know if the scripts they use guarantee that the preimage could not be longer than 32 bytes?

Yes. In fact [their Bitcoin script](https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#received-htlc-outputs) checks that the preimage is **exactly** 32 bytes, so 31 bytes is disallowed in Lightning.

Relevant snippet is:
```
OP_SIZE 32 OP_EQUAL
```

This means: "Add the size [of the preimage] to the stack, add 32 to the stack, check that the top two things on the stack are equal."

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 10]
```diff
@@ -1,11 +1,9 @@
 # Interledger Architecture
 
-This document outlines the Interledger architecture and explains how the different layers relate to each other.
+This document outlines the Interledger architecture and protocol suite and explains how the different layers relate to each other. The [Interledger Protocol (ILP)](#interledger-protocol-ilp) is one layer in the Interledger Architecture.
```
Comment (justmoon) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Saw the same wording in the index. Minor issue, but "architecture and protocol suite" reads weird to me, since the decision to split it into multiple protocol layers is just a feature of the architecture. So it's kinda like saying: "I like this car and sports car." Not really wrong, just a bit strange way to say it.

Like the clarifying sentence though on the relationship between ILP and Interledger architecture.

Would lower case "architecture".

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 11]
```diff
@@ -217,15 +229,24 @@ Application layer protocols deal with the exchange of payment details and associ
 
 The Simple Payment Setup Protocol (SPSP) is an application layer protocol for negotiating payment details. SPSP handles account and amount discovery, condition creation, quoting and setup. SPSP uses Webfinger ([RFC 7033](https://tools.ietf.org/html/rfc7033)) and an HTTP-based protocol for querying account and amount details, [ILQP](#interledger-quoting-protocol-ilqp) for quoting, and [ILP](#interledger-protocol-ilp) for payment execution.
 
-The protocol is described in [IL-RFC 9](../0009-simple-payment-setup-protocol/).
+The protocol is described in [IL-RFC 9](../0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md).
```
Comment (justmoon) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Note that we have to update the script that renders these specs for interledger.org. Currently, this link will link you to the unrendered `.md` file.

(But it is better to fix the link for Github users' benefit and then just rewrite it back as part of the rendering process for the website.)

File: 0003-interledger-protocol/0003-interledger-protocol.md [hunk 12]
```diff
@@ -172,78 +131,66 @@ As with the [internet protocol](https://tools.ietf.org/html/rfc791#section-2.3),
...
 ```
 
-Care must be taken in mapping interledger addresses to local ledger accounts. Examples of address mappings may be found in "Address Mappings" ((TODO)).
+More information about ILP addresses can be found in the [ILP Address Specification](../0015-ilp-addresses/).
+
+The mapping from addresses to local accounts on a ledger is defined by the ledger protocol.
 
 ### Connectors
 
 Connectors implement the interledger protocol to forward payments between ledgers. Connectors also implement the [Connector to Connector Protocol (CCP)](../0010-connector-to-connector-protocol/) to coordinate routing and other interledger control information.
```
Comment (mDuo13) [inline 0003-interledger-protocol/0003-interledger-protocol.md]: connectors also implement (or include a module that implements) the ledger protocol of the ledgers where they have accounts

File: 0003-interledger-protocol/0003-interledger-protocol.md [hunk 13]
```diff
@@ -172,78 +131,66 @@ As with the [internet protocol](https://tools.ietf.org/html/rfc791#section-2.3),
...
 
-![Interledger is an overlay across ledgers](assets/ilp-ledger-relation.png)
-
 ## Specification
 
-### ILP Header Format
+### ILP Payment Format
```
Comment (mDuo13) [inline 0003-interledger-protocol/0003-interledger-protocol.md]: Change this to "ILP Payment Packet Format"

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 14]
```diff
@@ -98,53 +98,62 @@ Ledger protocols are responsible for executing the individual transfers that con
...
 
-It MUST support a way to discover connectors connected to the ledger.
+The ledger MUST support a way to look up a fulfillment by condition hash. It SHOULD automatically reject new transfers (that have not been prepared yet) that have an execution condition for which the ledger already knows the fulfillment. This aids in [error recovery](#error-recovery).
 
-It MUST support a way to look up a fulfillment by condition hash. It SHOULD automatically reject new transfers (that have not been prepared yet) that have an execution condition for which the ledger already knows the fulfillment. This aids in [UTP error recovery](#error-recovery).
+The ledger MUST support preparing, executing, and rejecting transfers in 1 second or less.
+
+The ledger MUST define an [ILP Address](../0015-ilp-addresses/0015-ilp-addresses.md) prefix and scheme such that accounts on the ledger can be addressed using canonical ILP addresses.
+
+The ledger MUST support off-ledger, ledger-secured payment channels.
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: I think we should actually just take that out for now. The role of payment channels in ILP is something that we need to think more about and it's strange to call it a requirement for a ledger when we don't have a solid description of what exactly a ledger would need to do to fulfill that.

File: 0001-interledger-architecture/0001-interledger-architecture.md [hunk 15]
```diff
@@ -98,53 +98,62 @@ Ledger protocols are responsible for executing the individual transfers that con
...
-#### Simple Ledger Protocol (SLP)
+#### Five Bells Ledger Protocol (5BLP)
 
-Simple Ledger Protocol (SLP) is a RESTful, JSON-based protocol that was developed specifically to provide the minimum functionality required for full Interledger support.
+Five Bells Ledger Protocol (5BLP) is a RESTful, JSON-based protocol that was developed specifically to provide the minimum functionality required for full Interledger support.
 
-A reference implementation of a ledger using SLP can be found [here](https://github.com/interledger/five-bells-ledger).
+A reference implementation of a ledger using 5BLP can be found [here](https://github.com/interledger/five-bells-ledger).
 
 #### Blockchain Protocols (e.g. Bitcoin)
```
Comment (emschwartz) [inline 0001-interledger-architecture/0001-interledger-architecture.md]: Why? All the other things in that section have "(e.g. ...)" in the header
